### ISSUE TITLE
Empty string treated as missing message

### ISSUE THREAD

**Issue by bodhi** [yahoo/react-intl#607]
`formatMessage` treats a message of `""` (empty) string (and 0, and any other "falsy" value) the same as a missing message, resulting in:

```
[React Intl] Missing message: "<key>" for locale: "en-AU", using default message as fallback.
```

Which would be tolerable, except that:

* it will eventually fall back to the message ID, which is decidedly *not* blank.

* it logs this via `console.error`, which results in a full-screen red-box error message on React Native.

**Comment by ericf** [comment]
Yes, if a message is an empty string it's considered an error.

If you're missing translations for a locale, I'd recommend merging the translated messages with the default messages so that every key has a non-falsy value.

**Comment by bodhi** [comment]
> Yes, if a message is an empty string it's considered an error.

This means there is no way for a message to be localised to an empty string?

> If you're missing translations for a locale, I'd recommend merging the translated messages with the default messages so that every key has a non-falsy value.

The message isn't missing, it's intended to be blank. What I needed to do as a workaround is to use a string such as `" "` (a single space), and then `trim` it when it's used.

**Comment by ericf** [comment]
I don't follow; can you explain more?

**Comment by bodhi** [comment]
In our application, we have a set of "content cards" that may have a title, text and button label; and each card is identified by a key. Some of the cards need a title, but some do not. When we display the card, we can dynamically generate the message ID for e.g. the card title: `` `cards.${key}.title` ``.

If a localisation for a particular card doesn't require a title, then it could be set to the empty string and then the card component could skip rendering a blank title. But this doesn't work since `formatMessage` will fall back to the default message (which has the same problem), and then the message ID. So instead of an empty title, we'll get `"cards.cardId.title"`.

---

Basically my issue is that the system treats these as equivalent for a message id of `b`:

```
{
  a: "hello world"
}

{
  a: "hello world",
  b: null
}

{
  a: "hello world",
  b: ""
}
```

I understand that you want to bring attention to missing messages, and I've found the alert behaviour quite useful generally. But the "loss of meaning" created by treating all falsy values in the same way prevents (or rather, makes more difficult) my particular use case.

**Comment by ericf** [comment]
Ah okay. Check out my comment here: https://github.com/yahoo/react-intl/pull/488#issuecomment-224945705

You can use `injectIntl` to get access to the messages and check if it exists before trying to render it. You can use this in your Card component where the title message is optional.

**Comment by bodhi** [comment]
Aha! I tried searching various permutations of keywords for issues, but I guess I forgot to search PR comments too. Thanks, I'll look through that thread.

**Comment by greatbody** [comment]
in source of react-intj, src/format.js
line 203:
change:
if (message)
to:
if (Object.prototype.toString.apply(message) === '[object String]')
will solve the problem

**Comment by rasenplanscher** [comment]
@greatbody 
That will break when you update your package, which you should from time to time. And no, extracting from `node_modules` is usually not a good idea

**Comment by greatbody** [comment]
@rasenplanscher Yes, I do understand that.
To really solve this problem we need to change the code in repo.

**Comment by swernerx** [comment]
We just got an unexpected error in our application as well. One alternative solution might be to use a zero width space - this makes it very intentional that the translator wants an empty string: http://www.fileformat.info/info/unicode/char/200B/index.htm - not nice, but IMHO better than changing the implementation.

**Comment by asafna** [comment]
You can use <FormattedHTMLMessage /> and put &nbsp; for empty translations

**Comment by TrejGun** [comment]
Intentional empty strings should be supported

**Comment by JustFly1984** [comment]
I agree, there should be warning instead of error, at least for production. I need to wrap each

```javascript
if (value) {
  return intl.formatMessage({ id: value })
} else {
  return ''
}
```
in `React.useMemo()` pretty much everywhere in the codebase to escape app crashing. I've seen several crashes in Sentry logs in case server responds in unexpected way.

We have a goal to internationalize error and success messages, so server sends keys as error messages. but if there is unexpected error, or error is empty (false), we have crush in development or production, which is very unpleasant experience. And typescript does not help in this kind of cases.

Thank you for great work!

**Comment by TrejGun** [comment]
@ericf please reopen

**Comment by douira** [comment]
I'm internationalizing a react native app and have not found a good solution to formatting inside messages so I'm adding three messages. One before the value, one for the value itself and one after the value so that locales where the label comes before the value also work. This usually requires one of the messages to be empty.
Please reconsider and re-open this issue, you seem to have misunderstood when you closed it.

**Comment by longlho** [comment]
@douira can u provide an example?

**Comment by douira** [comment]
I'm formatting something like "**5** Posts" and in order to make the 5 bold there's some styled react native components involved. Now I can imagine there's a locale in which this is written "jfjgojlfgfd **5**" instead. I don't know such a language but there probably is one. If this doesn't happen in any language let me know and I'll simplify my code.

Here's an example:
```tsx
<Text style={styles.slimText}>
  <FormattedMessage
    id="ProfilePostCounter.posts.1-label"
    defaultMessage=""
    description="Posts counter pre-value label text."
    values={{ posts }}
  />
</Text>
<Text style={styles.boldText}>
  <FormattedMessage
    id="ProfilePostCounter.posts.2-value"
    defaultMessage="{posts, number}"
    description="Posts counter value text. This text is bold and should contain the number itself."
    values={{ posts }}
  />
</Text>
<Text style={styles.slimText}>
  <FormattedMessage
    id="ProfilePostCounter.posts.3-label"
    defaultMessage="{posts, plural, one {Beitrag} other {Beiträge}}"
    description="Posts counter post-value label text."
    values={{ posts }}
  />
</Text>
```
Is there a better way to solve this? I don't think I can use FormatJS' built-in rich text formatting since this is native and browser tags don't work as far as I know.

Edit: also wanted to say that my current workaround is using a zero width space like this ``defaultMessage={`\u200B`}`` but this is not a great solution since translations will also need to include this which most translations programs don't support.

**Comment by longlho** [comment]
@douira you should use our built-in rich text formatting. We use XML tag syntax but it's not related to browser at all. We run this in Node as well.

**Comment by isaachinman** [comment]
@ericf @longlho My team and I came across this (extremely) unexpected behaviour recently.

If content providers give an empty string or `null`, what they mean is that that locale should _not_ display the content.

I think this issue is a misunderstanding on the maintainers' part about the differences between falsy values in JavaScript.

An `undefined` value means exactly that – the value is not defined and can be treated as "missing". However, both empty strings and `null` values _are_ defined. They are intentional values provided by the user and should be respected instead of being treated the same as `undefined`.

This is a trivial fix. Are you open to at least providing this behaviour as a non-default, non-breaking, opt-in option? Otherwise we will be forced to either fork or move away from `react-intl`.

**Comment by longlho** [comment]
You're conflating what the message is to the logic of displaying the message. This library is in charge of formatting a message, and in the attempt to format the message if we encounter an error, we then trigger the fallback mechanism. `null`, `undefined`, `empty` are not format-able values. If you want to hide an empty message, then that should be part of your business logic. Hiding a message also causes unwanted side effects that is outside of our scope (like layout shifts).
Besides, TMSes also have very long SLA for translations (3days+ sometimes) and a lot of them either fill in untranslated messages with English string, or empty string.

**Comment by isaachinman** [comment]
> This library is in charge of formatting a message

Agreed. However, the easiest job a formatter can be assigned, is to do nothing. Is it the case that all string transformation packages/utils should error out on empty strings?

Besides that, which library are we talking about? I was referring to `react-intl`, not some lower level formatting lib. Perhaps the monorepo structure is creating confusion, apologies if so.

> If you want to hide an empty message, then that should be part of your business logic.

Is it the opinion of the maintainers that this type of behaviour cannot reasonably be expected of a frontend/React package? Writing business logic for a basic use case on top of `react-intl` and having to then write eslint rules to prevent devs from importing directly from `react-intl` is certainly less than ideal.

I'm wondering if a PR to introduce a non-breaking, opt-in option would be considered.

**Comment by TrejGun** [comment]
it is probably a good case for feature flags like
```jsx
<IntlProvider showWarningOnEmptyStrings><IntlProvider/>
```
or 
```jsx
<IntlProvider allowEmpty><IntlProvider/>
```

**Comment by longlho** [comment]
So the short answer is I'm ok w/ introducing an opt-in option for this if there's a PR. It should be called something like `shouldFallbackOnEmptyString` or such, and default to `true`.
The long answer is fallback mechanism within `react-intl` is an implementation detail that can change. Right now we only support single fallback chain (chosen `locale` -> `defaultLocale`) but it can get a lot more complicated with [locale negotiation](https://github.com/tc39/proposal-intl-localematcher) where you can fallback to multiple locales (e.g `zh-Hant-TW` -> `zh-Hant` -> `zh`). At the same time we don't really have this planned. In the future this might change.

**Comment by isaachinman** [comment]
> So the short answer is I'm ok w/ introducing an opt-in option for this if there's a PR

Great! I will speak with my team and hopefully get something forward very soon. Just to be clear, we would prefer if this option is at the provider (or similar) level, as opposed to having to define it per `FormattedMessage`/`useIntl` call. Does that seem acceptable?

> The long answer is fallback mechanism within react-intl is an implementation detail that can change

The implementation details of a locale fallback procedure are not in question here at all. It doesn't make a difference if our fallback logic has 2 locales or 20, the desired behaviour is to _not_ fallback on an empty string.

**Comment by dmendonez-mdsol** [comment]
> > Yes, if a message is an empty string it's considered an error.
> 
> This means there is no way for a message to be localised to an empty string?
> 
> > If you're missing translations for a locale, I'd recommend merging the translated messages with the default messages so that every key has a non-falsy value.
> 
> The message isn't missing, it's intended to be blank. What I needed to do as a workaround is to use a string such as `" "` (a single space), and then `trim` it when it's used.

The `trim` function works to get the intended result, although, not the ideal solution. Would just prefer that empty strings are supported. I would even say that `null` should be supported as well.

**Comment by jufemaiz** [comment]
> Right now we only support single fallback chain (chosen `locale` -> `defaultLocale`) but it can get a lot more complicated with [locale negotiation](https://github.com/tc39/proposal-intl-localematcher) where you can fallback to multiple locales (e.g `zh-Hant-TW` -> `zh-Hant` -> `zh`). At the same time we don't really have this planned. In the future this might change.

@longlho wondering if this has been considered further? Noting that, for example, Rails I18n has [support for structured fallbacks](https://guides.rubyonrails.org/configuring.html#config-i18n-fallbacks) since at least 2009.
